src/dh_lib.rs

```diff
@@ -0,0 +1,241 @@
+    paths_to_try.push(dir.join(format!("{}.{}", package, filename)));
+    paths_to_try.push(dir.join(named_filename.clone()));
+    paths_to_try.push(dir.join(filename.clone()));
+
+    for path_to_try in paths_to_try {
+        if path_to_try.is_file() {
+            return Some(path_to_try);
+        }
+    }
+
+    None
+}
+
+/// Build up one or more shell script fragments for a given maintainer script
+/// for a debian package in preparation for writing them into or as complete
+/// maintainer scripts in `apply()`, pulling fragments from a "library" of
+/// so-called "autoscripts".
+/// 
+/// Takes a map of values to search and replace in the selected "autoscript"
+/// fragment such as a systemd unit name placeholder and value.
+/// 
+/// # Cargo Deb specific behaviour
+/// 
+/// The autoscripts are sourced from within the binary via the rust_embed crate.
+/// 
+/// # Known limitations
+/// 
+/// Arbitrary sed command based file editing is not supported.
+/// 
+/// # References
+///
+/// https://git.launchpad.net/ubuntu/+source/debhelper/tree/lib/Debian/Debhelper/Dh_Lib.pm?h=applied/12.10ubuntu1#n1135
+pub(crate) fn autoscript(
+    tmp_dir: &Path,
+    package: &str,
+    script: &str,
+    snippet_filename: &str,
+    replacements: &HashMap<&str, String>,
+    listener: &mut dyn Listener)
+{
+    let bin_name = std::env::current_exe().unwrap();
+    let bin_name = bin_name.file_name().unwrap();
+    let bin_name = bin_name.to_str().unwrap();
+    let outfile = tmp_dir.join(format!("{}.{}.debhelper", package, script));
+
+    listener.info(format!("Maintainer script {} will be augmented with autoscript {}", &script, snippet_filename));
+
+    if outfile.exists() && (script == "postrm" || script == "prerm") {
+        if !replacements.is_empty() {
+            // prepend new text to existing file
+            let mut new_text = String::new();
+            new_text.push_str(&format!("# Automatically added by {}\n", bin_name));
+            new_text.push_str(&autoscript_sed(snippet_filename, replacements));
+            new_text.push_str("# End automatically added section\n");
+            new_text.push_str(&std::fs::read_to_string(&outfile).unwrap());
+            let mut file = File::create(&outfile).unwrap(); // deletes the original, if any
+            file.write_all(&new_text.into_bytes()).unwrap();
+        } else {
+            // We don't support sed commands yet.
+            unimplemented!();
+        }
+    } else if !replacements.is_empty() {
+        let mut new_text = String::new();
+        new_text.push_str(&format!("# Automatically added by {:?}\n", bin_name));
+        new_text.push_str(&autoscript_sed(snippet_filename, replacements));
+        new_text.push_str("# End automatically added section\n");
+        let mut file = if !outfile.exists() {
+            File::create(&outfile).unwrap()
+        } else {
+            OpenOptions::new().append(true).open(&outfile).unwrap()
```

Comment:
I considered that and agree with you. I didn't try that yet as I wanted first to get the debhelper dh_systemdinstall functionality working as it does in the original Perl implementation, then ensure I have tests passing that verify the working behaviour, then change the implementation to be in-memory based. I am happy to try updating it to work in-memory instead of with a temp folder.